library/psa_crypto_storage.h

```diff
@@ -59,7 +59,11 @@ extern "C" {
  * This limitation will probably become moot when we implement client
  * separation for key storage.
  */
+#if defined(MBEDTLS_PSA_CRYPTO_SPM)
+#define PSA_MAX_PERSISTENT_KEY_IDENTIFIER 0xffff000000000000
```

Comment:
I don't understand this value. Why is 0xffffffff00001234 ok (as the key id 0xffffffff with the owner 0x00001234)? What is this macro supposed to be used for? At the moment, it's one plus the maximum key id value that an application may use.

Reply:
This is your macro, we're not using it externally. This has been explained in https://github.com/ARMmbed/mbed-crypto/pull/32#discussion_r254597745
Please take a look

Reply:
But on PSA service integration, `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` does not limit the key id values that applications can use. Applications can use any value between 0 and 0xffffffff inclusive, unless there's a check in the service code. So what's the point of setting `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` to this particular value?

Reply:
The service uses the library as a black box. The service is not communicating with neither ITS (directly) nor any internal crypto APIs. So the code flow remains the same from the librarys' point of view.

Reply:
In this case, nothing is enforcing a particular range of key ids for an application. Which may be ok (it's a design decision that we're still free to take), but then `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` is not useful in a PSA service integration.

Reply:
Why not? The following code runs when crypto is used as a library but also as a PSA service:
https://github.com/ARMmbed/mbed-crypto/blob/2d7e5fe31da55d47748d259d11c08800261a4fd1/library/psa_crypto_slot_management.c#L205-L219

Shouldn't the library (in both modes) reserve key ids for it's internal use as stated in documentation? > Reserve a whole range of key slots just in case something else comes up.

Reply:
Ok, my bad, this _does_ enforce an upper limit of 0xfffeffff for key ids, since the key id is in the upper 32 bits. However, the checks as described are still not correct.

On an upgrade of ITS from 32-bit identifiers to 64-bit identifiers, how are identifiers converted?

* If the conversion is `uint64_t new_key_id = old_key_id`, then the entropy file is now `PSA_CRYPTO_ITS_RANDOM_SEED_UID` which is `0x00000000ffffff52`. This is key id 0 of owner 0xffffff52. Per https://github.com/ARMmbed/mbed-crypto/pull/32#discussion_r256532294 (no validation of key ids in the service) these are valid values. 0xffffff52 has the most significant bit set which makes it a non-secure partition. We don't have multiple non-secure partitions on Mbed OS (yet?), but we do on Linux.
* If the conversion is `uint64_t new_key_id = (uint64_t)old_key_id << 32`, then `PSA_CRYPTO_ITS_RANDOM_SEED_UID` needs to change to `0xffffff5200000000`, or we need an upgrade path.

Reply:
As far as I know the only key that must be preserved between Mbed OS versions 5.11 and 5.12 is the injected entropy key with id `PSA_CRYPTO_ITS_RANDOM_SEED_UID`. 

**_NON_-PSA Target: injected entropy preservation between versions 5.11 and 5.12:**
This has been tested and validated on a K64F (_non_-PSA target i.e. without the crypto service), see https://github.com/ARMmbed/mbed-crypto/pull/32#issuecomment-461457661.

**PSA Target: injected entropy preservation between versions 5.11 and 5.12:**
I don't see a reason to change the value of `PSA_CRYPTO_ITS_RANDOM_SEED_UID`, I'll try to explain my point of view...

The following is the entropy inject function implemented in mbed crypto:
https://github.com/ARMmbed/mbed-crypto/blob/0574e6a7bd0bec47e45bbd30db2e5236896b38ea/library/psa_crypto.c#L4427-L4455
As you can see entropy injection does _not_ go through the "normal" flow other crypto keys do (i.e. `psa_open_key` or `psa_create_key` etc.), it talks to the ITS directly. There's no intervention by the crypto service (or anything else) that changes this key id into anything else. This means that we can keep using the same `PSA_CRYPTO_ITS_RANDOM_SEED_UID` value both in a system where ITS ids are 32 or 64 bits. The ITS does not care about the _meaning_ of the id it recieves.

As to overwriting or obtaining a handle to a key with id `PSA_CRYPTO_ITS_RANDOM_SEED_UID`, the only way to access this ITS record is by talking to ITS directly, it will not be possible to do so via crypto API:

**In a PSA target**, if a user were to call `psa_open_key` or `psa_create_key` with id `PSA_CRYPTO_ITS_RANDOM_SEED_UID` the service would translate this value to:
* If calling from NSPE (partition id `-1`) - ID sent to ITS = `0xffffff52ffffffff`
* If calling from different partition (for example partition id `3`) - ID sent to ITS = `0xffffff5200000003`

So I don't see a possible way of obtaining a handle to a key with id `0xffffff52` when going through the crypto service.

Moreover, the results from these 2 examples  would not pass the check (in function `psa_internal_make_key_persistent`) against `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` (when it's defined to `0xffff000000000000`) as both `0xffffff52ffffffff` and `0xffffff5200000003` are greater than `0xffff000000000000` and the function would return a `PSA_ERROR_INVALID_ARGUMENT`.

**In a non-PSA system**, the above is also true, if `PSA_MAX_PERSISTENT_KEY_IDENTIFIER` is defined as `0xffff0000` (as it is now), it will also fail the validation in `psa_internal_make_key_persistent` as `0xffffff52` is greater than `0xffff0000`.

One more point, the crypto service does (as part of https://github.com/ARMmbed/mbed-os/pull/9575) validate that key id is **not** 0 _prior_ to calling the library's open/create a key, see:
https://github.com/ARMmbed/mbed-os/blob/3bc5f9086c3bbf1471906d6a4753457cf46f19cf/components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c#L1179
and
https://github.com/ARMmbed/mbed-os/blob/3bc5f9086c3bbf1471906d6a4753457cf46f19cf/components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c#L1201

Clarification on:
> the only way to access this ITS record is by talking to ITS directly

It will only be possible to do so from the partition that created the ITS record. ITS has it's own validation/enforcement mechanism which prevents accessing ITS records from partitions which are not the record's owner. They do so by converting the id received into a _string_ and prepending the calling partition id to that string. So if ITS was called **directly** from NSPE (`0xffffffff`) with id `PSA_CRYPTO_ITS_RANDOM_SEED_UID` (`0xffffff52`), the actual ITS record id would be converted (by ITS) into `ffffffff_ffffff52`. In our case, in a PSA system (where the crypto service id is `0x23`), the injected entropy id that **ITS** eventually stores (as a string) is `23_ffffff52`. This ensures that the entropy key will only be accessible from the crypto partition and only via direct access to ITS (i.e. without going through open/create key APIs).

Reply:
> **In a PSA target**, if a user were to call `psa_open_key` or `psa_create_key` with id `PSA_CRYPTO_ITS_RANDOM_SEED_UID` the service would translate this value to:

> If calling from NSPE (partition id `-1`) - ID sent to ITS = `0xffffff52ffffffff`

On Linux, there are multiple partitions in the NSPE and they're numbered 0xffffffff, 0xfffffffe, etc. I don't know if it's already implemented, but if it isn't it's only a matter of weeks, and we need a solution that works in this time scale.

If partition 0xffffff52 wants to store key ID 0, this can only be blocked at the service level. The safety of the entropy seed file depends on this check in the service. I would strongly prefer a more robust solution where the entropy seed is in a domain that the PRoT already reserves for itself: owner_uid = 0.